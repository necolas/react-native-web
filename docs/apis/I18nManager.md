# I18nManager

Control and set the layout and writing direction of the application. You must
set `dir="rtl"` (and should set `lang="${lang}"`) on the root element of your
app.

## Properties

**isRTL**: bool = false

Whether the application is currently in RTL mode.

## Methods

static **allowRTL**(allowRTL: bool)

Allow the application to display in RTL mode.

static **forceRTL**(allowRTL: bool)

Force the application to display in RTL mode.

static **setRTL**(allowRTL: bool)

Set the application to display in RTL mode. You will need to determine the
user's preferred locale and if it is an RTL language. (This is best done on the
server as it is notoriously inaccurate to deduce client-side.)
