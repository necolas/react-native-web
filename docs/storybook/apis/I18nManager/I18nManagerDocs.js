/* eslint-disable react/jsx-sort-props */

/**
 * @flow
 */

import RTLToggleExample from './examples/RTLToggle';
import React from 'react';
import { storiesOf } from '@kadira/storybook';
import UIExplorer, { DocItem } from '../../ui-explorer';

const sections = [
  {
    title: 'Properties',
    entries: [
      <DocItem
        name="isRTL"
        typeInfo="boolean = false"
        description="Whether the application is currently in RTL mode."
      />
    ]
  },
  {
    title: 'Methods',
    entries: [
      <DocItem
        name="static allowRTL"
        typeInfo="(allowRTL: boolean) => void"
        description="Allow the application to display in RTL mode."
      />,

      <DocItem
        name="static forceRTL"
        typeInfo="(forceRTL: boolean) => void"
        description="Force the application to display in RTL mode."
      />,

      <DocItem
        label="web"
        name="static setPreferredLanguageRTL"
        typeInfo="(isRTL: boolean) => void"
        description="Set the application's preferred writing direction to RTL. You will need to determine the user's preferred locale server-side (from HTTP headers) and decide whether it's an RTL language."
      />
    ]
  },
  {
    title: 'Examples',
    entries: [
      <DocItem
        description="Toggling LTR/RTL layout at runtime"
        example={{
          render: () => <RTLToggleExample />
        }}
      />
    ]
  }
];

storiesOf('APIs', module).add('I18nManager', () =>
  <UIExplorer
    description="Control and set the layout and writing direction of the application."
    sections={sections}
    title="I18nManager"
    url="apis/I18nManager"
  />
);
