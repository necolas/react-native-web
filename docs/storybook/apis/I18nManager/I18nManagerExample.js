import { storiesOf } from '@kadira/storybook';
import UIExplorer from '../../UIExplorer';
import { I18nManager, StyleSheet, TouchableHighlight, Text, View } from 'react-native';
import React, { Component } from 'react';

class I18nManagerExample extends Component {
  componentWillUnmount() {
    I18nManager.setPreferredLanguageRTL(false);
  }

  render() {
    return (
      <View style={styles.container}>
        <Text accessibilityRole="heading" style={styles.welcome}>
          LTR/RTL layout example!
        </Text>
        <Text style={styles.text}>
          The writing direction of text is automatically determined by the browser, independent of
          the global writing direction of the app.
        </Text>
        <Text style={[styles.text, styles.rtlText]}>
          أحب اللغة العربية
        </Text>
        <Text style={[styles.text, styles.textAlign]}>
          textAlign toggles
        </Text>
        <View style={styles.horizontal}>
          <View style={[styles.box, { backgroundColor: 'lightblue' }]}>
            <Text>One</Text>
          </View>
          <View style={[styles.box]}>
            <Text>Two</Text>
          </View>
        </View>
        <TouchableHighlight
          onPress={this._handleToggle}
          style={styles.toggle}
          underlayColor="rgba(0,0,0,0.25)"
        >
          <Text>Toggle LTR/RTL</Text>
        </TouchableHighlight>
      </View>
    );
  }

  _handleToggle = () => {
    I18nManager.setPreferredLanguageRTL(!I18nManager.isRTL);
    this.forceUpdate();
  };
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#F5FCFF',
    flex: 1,
    justifyContent: 'center',
    padding: 10
  },
  welcome: {
    fontSize: 28,
    marginVertical: 10
  },
  text: {
    color: '#333333',
    fontSize: 18,
    marginBottom: 5
  },
  textAlign: {
    textAlign: 'left'
  },
  horizontal: {
    flexDirection: 'row',
    marginVertical: 10
  },
  box: {
    borderWidth: 1,
    flex: 1
  },
  toggle: {
    alignSelf: 'center',
    borderColor: 'black',
    borderStyle: 'solid',
    borderWidth: 1,
    marginTop: 10,
    padding: 10
  }
});

const examples = [
  {
    title: 'RTL toggle',
    render: () => <I18nManagerExample />
  }
];

storiesOf('APIs', module).add('I18nManager', () =>
  <UIExplorer examples={examples} title="I18nManager" />
);
