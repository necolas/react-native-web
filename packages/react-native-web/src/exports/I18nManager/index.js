/**
 * Copyright (c) 2016-present, Nicolas Gallagher.
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @providesModule I18nManager
 * @flow
 */

import ExecutionEnvironment from 'fbjs/lib/ExecutionEnvironment';

type I18nManagerStatus = {
  allowRTL: (allowRTL: boolean) => void,
  doLeftAndRightSwapInRTL: boolean,
  forceRTL: (forceRTL: boolean) => void,
  isRTL: boolean,
  setPreferredLanguageRTL: (setRTL: boolean) => void,
  swapLeftAndRightInRTL: (flipStyles: boolean) => void
};

let doLeftAndRightSwapInRTL = true;
let isPreferredLanguageRTL = false;
let isRTLAllowed = true;
let isRTLForced = false;

const isRTL = () => {
  if (isRTLForced) {
    return true;
  }
  return isRTLAllowed && isPreferredLanguageRTL;
};

const onDirectionChange = () => {
  if (ExecutionEnvironment.canUseDOM) {
    if (document.documentElement && document.documentElement.setAttribute) {
      document.documentElement.setAttribute('dir', isRTL() ? 'rtl' : 'ltr');
    }
  }
};

const I18nManager: I18nManagerStatus = {
  allowRTL(bool) {
    isRTLAllowed = bool;
    onDirectionChange();
  },
  forceRTL(bool) {
    isRTLForced = bool;
    onDirectionChange();
  },
  setPreferredLanguageRTL(bool) {
    isPreferredLanguageRTL = bool;
    onDirectionChange();
  },
  swapLeftAndRightInRTL(bool) {
    doLeftAndRightSwapInRTL = bool;
  },
  get doLeftAndRightSwapInRTL() {
    return doLeftAndRightSwapInRTL;
  },
  get isRTL() {
    return isRTL();
  }
};

export default I18nManager;
