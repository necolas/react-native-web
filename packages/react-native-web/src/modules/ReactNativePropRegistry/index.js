/**
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @providesModule ReactNativePropRegistry
 * @flow
 */

const emptyObject = {};
const objects = {};
const prefix = 'r';
let uniqueID = 1;

const createKey = id => `${prefix}-${id}`;

export default class ReactNativePropRegistry {
  static register(object: Object): number {
    const id = uniqueID++;
    if (process.env.NODE_ENV !== 'production') {
      Object.freeze(object);
    }
    const key = createKey(id);
    objects[key] = object;
    return id;
  }

  static getByID(id: number): Object {
    if (!id) {
      // Used in the style={[condition && id]} pattern,
      // we want it to be a no-op when the value is false or null
      return emptyObject;
    }
    const key = createKey(id);
    const object = objects[key];
    if (!object) {
      console.warn('Invalid style with id `' + id + '`. Skipping ...');
      return emptyObject;
    }
    return object;
  }
}
