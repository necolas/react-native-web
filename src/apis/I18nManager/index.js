/**
 * Copyright (c) 2016-present, Nicolas Gallagher.
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @providesModule I18nManager
 * @flow
 */

import ExecutionEnvironment from 'fbjs/lib/ExecutionEnvironment';

type I18nManagerStatus = {
  allowRTL: (allowRTL: boolean) => void,
  forceRTL: (forceRTL: boolean) => void,
  setPreferredLanguageRTL: (setRTL: boolean) => void,
  isRTL: boolean
};

let isPreferredLanguageRTL = false;
let isRTLAllowed = true;
let isRTLForced = false;

const isRTL = () => {
  if (isRTLForced) {
    return true;
  }
  return isRTLAllowed && isPreferredLanguageRTL;
};

const onChange = () => {
  if (ExecutionEnvironment.canUseDOM) {
    if (document.documentElement && document.documentElement.setAttribute) {
      document.documentElement.setAttribute('dir', isRTL() ? 'rtl' : 'ltr');
    }
  }
};

const I18nManager: I18nManagerStatus = {
  allowRTL(bool) {
    isRTLAllowed = bool;
    onChange();
  },
  forceRTL(bool) {
    isRTLForced = bool;
    onChange();
  },
  setPreferredLanguageRTL(bool) {
    isPreferredLanguageRTL = bool;
    onChange();
  },
  get isRTL() {
    return isRTL();
  }
};

export default I18nManager;
