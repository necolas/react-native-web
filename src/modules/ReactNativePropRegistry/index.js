/* eslint-disable */
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @providesModule ReactNativePropRegistry
 * @flow
 */
'use strict';

const emptyObject = {};
const objects = {};
const prefix = 'r';
let uniqueID = 1;

const createKey = (id) => `${prefix}-${id}`;

class ReactNativePropRegistry {
  static register(object: Object): number {
    let id = uniqueID++;
    if (process.env.NODE_ENV !== 'production') {
      Object.freeze(object);
    }
    const key = createKey(id);
    objects[key] = object;
    return id;
  }

  static getByID(id: number): Object {
    if (!id) {
      // Used in the style={[condition && id]} pattern,
      // we want it to be a no-op when the value is false or null
      return emptyObject;
    }
    const key = createKey(id);
    const object = objects[key];
    if (!object) {
      console.warn('Invalid style with id `' + id + '`. Skipping ...');
      return emptyObject;
    }
    return object;
  }
}

module.exports = ReactNativePropRegistry;
