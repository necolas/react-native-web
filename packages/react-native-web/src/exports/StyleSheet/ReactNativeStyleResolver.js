/**
 * Copyright (c) 2016-present, Nicolas Gallagher.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @noflow
 */

/**
 * WARNING: changes to this file in particular can cause significant changes to
 * the results of render performance benchmarks.
 */

import { canUseDOM } from 'fbjs/lib/ExecutionEnvironment';
import createReactDOMStyle from './createReactDOMStyle';
import flattenArray from '../../modules/flattenArray';
import flattenStyle from './flattenStyle';
import I18nManager from '../I18nManager';
import i18nStyle from './i18nStyle';
import { prefixInlineStyles } from '../../modules/prefixStyles';
import StyleSheetManager from './StyleSheetManager';

const emptyObject = {};

export default class ReactNativeStyleResolver {
  _init() {
    this.cache = { ltr: {}, rtl: {}, rtlNoSwap: {} };
    this.injectedCache = { ltr: {}, rtl: {}, rtlNoSwap: {} };
    this.styleSheetManager = new StyleSheetManager();
  }

  constructor() {
    this._init();
  }

  getStyleSheet() {
    // reset state on the server so critical css is always the result
    const sheet = this.styleSheetManager.getStyleSheet();
    if (!canUseDOM) {
      this._init();
    }
    return sheet;
  }

  _injectRegisteredStyle(id) {
    const { doLeftAndRightSwapInRTL, isRTL } = I18nManager;
    const dir = isRTL ? (doLeftAndRightSwapInRTL ? 'rtl' : 'rtlNoSwap') : 'ltr';
    if (!this.injectedCache[dir][id]) {
      const style = flattenStyle(id);
      const domStyle = createReactDOMStyle(i18nStyle(style));
      Object.keys(domStyle).forEach(styleProp => {
        const value = domStyle[styleProp];
        if (value != null) {
          this.styleSheetManager.injectDeclaration(styleProp, value);
        }
      });
      this.injectedCache[dir][id] = true;
    }
  }

  /**
   * Resolves a React Native style object to DOM attributes
   */
  resolve(style) {
    if (!style) {
      return emptyObject;
    }

    // fast and cachable
    if (typeof style === 'number') {
      this._injectRegisteredStyle(style);
      const key = createCacheKey(style);
      return this._resolveStyleIfNeeded(style, key);
    }

    // resolve a plain RN style object
    if (!Array.isArray(style)) {
      return this._resolveStyleIfNeeded(style);
    }

    // flatten the style array
    // cache resolved props when all styles are registered
    // otherwise fallback to resolving
    const flatArray = flattenArray(style);
    let isArrayOfNumbers = true;
    let cacheKey = '';
    for (let i = 0; i < flatArray.length; i++) {
      const id = flatArray[i];
      if (typeof id !== 'number') {
        isArrayOfNumbers = false;
      } else {
        if (isArrayOfNumbers) {
          cacheKey += (id + '-');
        }
        this._injectRegisteredStyle(id);
      }
    }
    const key = isArrayOfNumbers ? createCacheKey(cacheKey) : null;
    return this._resolveStyleIfNeeded(flatArray, key);
  }

  /**
   * Resolves a React Native style object to DOM attributes, accounting for
   * the existing styles applied to the DOM node.
   *
   * To determine the next style, some of the existing DOM state must be
   * converted back into React Native styles.
   */
  resolveWithNode(rnStyleNext, node) {
    const { classList: rdomClassList, style: rdomStyle } = getDOMStyleInfo(node);
    // Convert the DOM classList back into a React Native form
    // Preserves unrecognized class names.
    const { classList: rnClassList, style: rnStyle } = rdomClassList.reduce(
      (styleProps, className) => {
        const { prop, value } = this.styleSheetManager.getDeclaration(className);
        if (prop) {
          styleProps.style[prop] = value;
        } else {
          styleProps.classList.push(className);
        }
        return styleProps;
      },
      { classList: [], style: {} }
    );

    // Create next DOM style props from current and next RN styles
    const { classList: rdomClassListNext, style: rdomStyleNext } = this.resolve([
      i18nStyle(rnStyle),
      rnStyleNext
    ]);

    // Final className
    // Add the current class names not managed by React Native
    const className = classListToString(rdomClassListNext.concat(rnClassList));

    // Final style
    // Next class names take priority over current inline styles
    const style = { ...rdomStyle };
    rdomClassListNext.forEach(className => {
      const { prop } = this.styleSheetManager.getDeclaration(className);
      if (style[prop]) {
        style[prop] = '';
      }
    });
    // Next inline styles take priority over current inline styles
    Object.assign(style, rdomStyleNext);

    return { className, style };
  }

  /**
   * Resolves a React Native style object
   */
  _resolveStyle(style) {
    const flatStyle = flattenStyle(style);
    const domStyle = createReactDOMStyle(i18nStyle(flatStyle));

    const props = Object.keys(domStyle).reduce(
      (props, styleProp) => {
        const value = domStyle[styleProp];
        if (value != null) {
          const className = this.styleSheetManager.getClassName(styleProp, value);
          if (className) {
            props.classList.push(className);
          } else {
            // Certain properties and values are not transformed by 'createReactDOMStyle' as they
            // require more complex transforms into multiple CSS rules. Here we assume that StyleManager
            // can bind these styles to a className, and prevent them becoming invalid inline-styles.
            if (
              styleProp === 'pointerEvents' ||
              styleProp === 'placeholderTextColor' ||
              styleProp === 'animationName'
            ) {
              const className = this.styleSheetManager.injectDeclaration(styleProp, value);
              if (className) {
                props.classList.push(className);
              }
            } else {
              if (!props.style) {
                props.style = {};
              }
              // 4x slower render
              props.style[styleProp] = value;
            }
          }
        }
        return props;
      },
      { classList: [] }
    );

    props.className = classListToString(props.classList);
    if (props.style) {
      props.style = prefixInlineStyles(props.style);
    }
    return props;
  }

  /**
   * Caching layer over 'resolveStyle'
   */
  _resolveStyleIfNeeded(style, key) {
    if (key) {
      const { doLeftAndRightSwapInRTL, isRTL } = I18nManager;
      const dir = isRTL ? (doLeftAndRightSwapInRTL ? 'rtl' : 'rtlNoSwap') : 'ltr';
      if (!this.cache[dir][key]) {
        // slow: convert style object to props and cache
        this.cache[dir][key] = this._resolveStyle(style);
      }
      return this.cache[dir][key];
    }
    return this._resolveStyle(style);
  }
}

/**
 * Misc helpers
 */
const createCacheKey = id => {
  const prefix = 'rn';
  return `${prefix}-${id}`;
};

const classListToString = list => list.join(' ').trim();

/**
 * Copies classList and style data from a DOM node
 */
const hyphenPattern = /-([a-z])/g;
const toCamelCase = str => str.replace(hyphenPattern, m => m[1].toUpperCase());

const getDOMStyleInfo = node => {
  const nodeStyle = node.style;
  const classList = Array.prototype.slice.call(node.classList);
  const style = {};
  // DOM style is a CSSStyleDeclaration
  // https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration
  for (let i = 0; i < nodeStyle.length; i += 1) {
    const property = nodeStyle.item(i);
    if (property) {
      // DOM style uses hyphenated prop names and may include vendor prefixes
      // Transform back into React DOM style.
      style[toCamelCase(property)] = nodeStyle.getPropertyValue(property);
    }
  }
  return { classList, style };
};
