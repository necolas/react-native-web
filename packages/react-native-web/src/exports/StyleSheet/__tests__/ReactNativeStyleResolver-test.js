/* eslint-env jasmine, jest */

import I18nManager from '../../I18nManager';
import ReactNativePropRegistry from '../../../modules/ReactNativePropRegistry';
import ReactNativeStyleResolver from '../ReactNativeStyleResolver';

let styleResolver;

describe('StyleSheet/ReactNativeStyleResolver', () => {
  beforeEach(() => {
    styleResolver = new ReactNativeStyleResolver();
  });

  describe('resolve', () => {
    const styleA = { borderWidth: 0, borderColor: 'red', width: 100 };
    const styleB = {
      position: 'absolute',
      left: 50,
      opacity: 0.5
    };
    const styleC = { width: 200 };

    const testResolve = (a, b, c) => {
      // no common properties, different resolving order, same result
      const resolve1 = styleResolver.resolve([a, b]);
      expect(resolve1).toMatchSnapshot();
      const resolve2 = styleResolver.resolve([b, a]);
      expect(resolve1).toEqual(resolve2);

      // common properties, different resolving order, different result
      const resolve3 = styleResolver.resolve([a, b, c]);
      expect(resolve3).toMatchSnapshot();
      const resolve4 = styleResolver.resolve([c, a, b]);
      expect(resolve4).toMatchSnapshot();
      expect(resolve3).not.toEqual(resolve4);
    };

    test('with register, resolves to className', () => {
      const a = ReactNativePropRegistry.register(styleA);
      const b = ReactNativePropRegistry.register(styleB);
      const c = ReactNativePropRegistry.register(styleC);
      testResolve(a, b, c);
    });

    test('with register before RTL, resolves to correct className', () => {
      const a = ReactNativePropRegistry.register({ left: '12.34%' });
      const b = ReactNativePropRegistry.register({ textAlign: 'left' });
      const c = ReactNativePropRegistry.register({ marginLeft: 10 });
      I18nManager.forceRTL(true);

      const resolved1 = styleResolver.resolve([a, b, c]);
      expect(resolved1).toMatchSnapshot();

      I18nManager.swapLeftAndRightInRTL(false);

      const resolved2 = styleResolver.resolve([a, b, c]);
      expect(resolved2).toMatchSnapshot();

      I18nManager.swapLeftAndRightInRTL(true);
      I18nManager.forceRTL(false);
    });

    test('with register, resolves to mixed', () => {
      const a = styleA;
      const b = ReactNativePropRegistry.register(styleB);
      const c = ReactNativePropRegistry.register(styleC);
      testResolve(a, b, c);
    });

    test('without register, resolves to inline styles', () => {
      testResolve(styleA, styleB, styleC);
    });

    test('resolves inline-style pointerEvents to classname', () => {
      expect(styleResolver.resolve({ pointerEvents: 'box-none' })).toMatchSnapshot();
    });
  });

  describe('resolveWithNode', () => {
    let node;

    beforeEach(() => {
      node = document.createElement('div');
    });

    test('preserves unrelated class names', () => {
      node.classList.add('unknown-class-1', 'unknown-class-2');
      const resolved = styleResolver.resolveWithNode({}, node);
      expect(resolved).toMatchSnapshot();
    });

    test('preserves unrelated inline styles', () => {
      node.style.cssText = 'font-size: 20px;';
      const resolved = styleResolver.resolveWithNode({ opacity: 1 }, node);
      expect(resolved).toMatchSnapshot();
    });

    test('next class names have priority over current inline styles', () => {
      node.style.cssText = 'opacity: 0.5;';
      const nextStyle = ReactNativePropRegistry.register({ opacity: 1 });
      const resolved = styleResolver.resolveWithNode(nextStyle, node);
      expect(resolved).toMatchSnapshot();
    });

    test('next inline styles have priority over current inline styles', () => {
      // note: this also checks for correctly uppercasing the first letter of DOM vendor prefixes
      node.style.cssText = 'opacity: 0.5; transform: scale(1);';
      const style = { opacity: 1, transform: [{ scale: 2 }] };
      const resolved = styleResolver.resolveWithNode(style, node);
      expect(resolved).toMatchSnapshot();
    });

    test('when isRTL=true, resolves to flipped inline styles', () => {
      // note: DOM state resolved from { marginLeft: 5, left: 5 } in RTL mode
      node.style.cssText = 'margin-right: 5px; right: 5px;';
      I18nManager.forceRTL(true);
      const resolved = styleResolver.resolveWithNode({ marginLeft: 10, right: 10 }, node);
      I18nManager.forceRTL(false);
      expect(resolved).toMatchSnapshot();
    });

    test('when isRTL=true, resolves to flipped classNames', () => {
      // note: DOM state resolved from { marginLeft: 5, left: 5 }
      node.style.cssText = 'margin-right: 5px; right: 5px;';
      const nextStyle = ReactNativePropRegistry.register({ marginLeft: 10, right: 1 });

      I18nManager.forceRTL(true);
      const resolved = styleResolver.resolveWithNode(nextStyle, node);
      I18nManager.forceRTL(false);
      expect(resolved).toMatchSnapshot();
    });

    test('when isRTL=true & doLeftAndRightSwapInRTL=false, resolves to non-flipped inline styles', () => {
      // note: DOM state resolved from { marginRight 5, right: 5, paddingEnd: 5 }
      node.style.cssText = 'margin-right: 5px; right: 5px; padding-left: 5px';
      I18nManager.forceRTL(true);
      I18nManager.swapLeftAndRightInRTL(false);
      const resolved = styleResolver.resolveWithNode(
        { marginRight: 10, right: 10, paddingEnd: 10 },
        node
      );
      I18nManager.forceRTL(false);
      I18nManager.swapLeftAndRightInRTL(true);
      expect(resolved).toMatchSnapshot();
    });
  });
});
