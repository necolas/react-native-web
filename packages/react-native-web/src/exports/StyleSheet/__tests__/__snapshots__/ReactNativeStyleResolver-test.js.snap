// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleSheet/ReactNativeStyleResolver resolve resolves inline-style pointerEvents to classname 1`] = `
Object {
  "classList": Array [
    "rn-pointerEvents-12vffkv",
  ],
  "className": "rn-pointerEvents-12vffkv",
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolve with register before RTL, resolves to correct className 1`] = `
Object {
  "classList": Array [
    "rn-marginRight-zso239",
    "rn-right-1bnbe1j",
    "rn-textAlign-1ff274t",
  ],
  "className": "rn-marginRight-zso239 rn-right-1bnbe1j rn-textAlign-1ff274t",
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolve with register before RTL, resolves to correct className 2`] = `
Object {
  "classList": Array [
    "rn-left-2s0hu9",
    "rn-marginLeft-1n0xq6e",
    "rn-textAlign-fdjqy7",
  ],
  "className": "rn-left-2s0hu9 rn-marginLeft-1n0xq6e rn-textAlign-fdjqy7",
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolve with register, resolves to className 1`] = `
Object {
  "classList": Array [
    "rn-borderTopColor-1gxhl28",
    "rn-borderRightColor-knoah9",
    "rn-borderBottomColor-1ani3fp",
    "rn-borderLeftColor-ribj9x",
    "rn-borderTopWidth-13yce4e",
    "rn-borderRightWidth-fnigne",
    "rn-borderBottomWidth-ndvcnb",
    "rn-borderLeftWidth-gxnn5r",
    "rn-left-1tsx3h3",
    "rn-opacity-icoktb",
    "rn-position-u8s1d",
    "rn-width-b8lwoo",
  ],
  "className": "rn-borderTopColor-1gxhl28 rn-borderRightColor-knoah9 rn-borderBottomColor-1ani3fp rn-borderLeftColor-ribj9x rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-left-1tsx3h3 rn-opacity-icoktb rn-position-u8s1d rn-width-b8lwoo",
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolve with register, resolves to className 2`] = `
Object {
  "classList": Array [
    "rn-borderTopColor-1gxhl28",
    "rn-borderRightColor-knoah9",
    "rn-borderBottomColor-1ani3fp",
    "rn-borderLeftColor-ribj9x",
    "rn-borderTopWidth-13yce4e",
    "rn-borderRightWidth-fnigne",
    "rn-borderBottomWidth-ndvcnb",
    "rn-borderLeftWidth-gxnn5r",
    "rn-left-1tsx3h3",
    "rn-opacity-icoktb",
    "rn-position-u8s1d",
    "rn-width-l0gwng",
  ],
  "className": "rn-borderTopColor-1gxhl28 rn-borderRightColor-knoah9 rn-borderBottomColor-1ani3fp rn-borderLeftColor-ribj9x rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-left-1tsx3h3 rn-opacity-icoktb rn-position-u8s1d rn-width-l0gwng",
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolve with register, resolves to className 3`] = `
Object {
  "classList": Array [
    "rn-borderTopColor-1gxhl28",
    "rn-borderRightColor-knoah9",
    "rn-borderBottomColor-1ani3fp",
    "rn-borderLeftColor-ribj9x",
    "rn-borderTopWidth-13yce4e",
    "rn-borderRightWidth-fnigne",
    "rn-borderBottomWidth-ndvcnb",
    "rn-borderLeftWidth-gxnn5r",
    "rn-left-1tsx3h3",
    "rn-opacity-icoktb",
    "rn-position-u8s1d",
    "rn-width-b8lwoo",
  ],
  "className": "rn-borderTopColor-1gxhl28 rn-borderRightColor-knoah9 rn-borderBottomColor-1ani3fp rn-borderLeftColor-ribj9x rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-left-1tsx3h3 rn-opacity-icoktb rn-position-u8s1d rn-width-b8lwoo",
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolve with register, resolves to mixed 1`] = `
Object {
  "classList": Array [
    "rn-left-1tsx3h3",
    "rn-opacity-icoktb",
    "rn-position-u8s1d",
  ],
  "className": "rn-left-1tsx3h3 rn-opacity-icoktb rn-position-u8s1d",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "width": "100px",
  },
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolve with register, resolves to mixed 2`] = `
Object {
  "classList": Array [
    "rn-left-1tsx3h3",
    "rn-opacity-icoktb",
    "rn-position-u8s1d",
    "rn-width-l0gwng",
  ],
  "className": "rn-left-1tsx3h3 rn-opacity-icoktb rn-position-u8s1d rn-width-l0gwng",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
  },
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolve with register, resolves to mixed 3`] = `
Object {
  "classList": Array [
    "rn-left-1tsx3h3",
    "rn-opacity-icoktb",
    "rn-position-u8s1d",
  ],
  "className": "rn-left-1tsx3h3 rn-opacity-icoktb rn-position-u8s1d",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "width": "100px",
  },
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolve without register, resolves to inline styles 1`] = `
Object {
  "classList": Array [],
  "className": "",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "left": "50px",
    "opacity": 0.5,
    "position": "absolute",
    "width": "100px",
  },
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolve without register, resolves to inline styles 2`] = `
Object {
  "classList": Array [],
  "className": "",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "left": "50px",
    "opacity": 0.5,
    "position": "absolute",
    "width": "200px",
  },
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolve without register, resolves to inline styles 3`] = `
Object {
  "classList": Array [],
  "className": "",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "left": "50px",
    "opacity": 0.5,
    "position": "absolute",
    "width": "100px",
  },
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolveWithNode next class names have priority over current inline styles 1`] = `
Object {
  "className": "rn-opacity-6dt33c",
  "style": Object {
    "opacity": "",
  },
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolveWithNode next inline styles have priority over current inline styles 1`] = `
Object {
  "className": "",
  "style": Object {
    "WebkitTransform": "scale(2)",
    "opacity": 1,
    "transform": "scale(2)",
  },
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolveWithNode preserves unrelated class names 1`] = `
Object {
  "className": "unknown-class-1 unknown-class-2",
  "style": Object {},
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolveWithNode preserves unrelated inline styles 1`] = `
Object {
  "className": "",
  "style": Object {
    "fontSize": "20px",
    "opacity": 1,
  },
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolveWithNode when isRTL=true & doLeftAndRightSwapInRTL=false, resolves to non-flipped inline styles 1`] = `
Object {
  "className": "",
  "style": Object {
    "marginRight": "10px",
    "paddingLeft": "10px",
    "right": "10px",
  },
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolveWithNode when isRTL=true, resolves to flipped classNames 1`] = `
Object {
  "className": "rn-left-1u10d71 rn-marginRight-zso239",
  "style": Object {
    "marginRight": "",
    "right": "5px",
  },
}
`;

exports[`StyleSheet/ReactNativeStyleResolver resolveWithNode when isRTL=true, resolves to flipped inline styles 1`] = `
Object {
  "className": "",
  "style": Object {
    "left": "10px",
    "marginRight": "10px",
    "right": "5px",
  },
}
`;
